### DISCUSSION TITLE
Keycloak SPI using Quarkus

### DISCUSSION THREAD

**Discussion by akil-rails** [quarkusio/quarkus#41056] [Community]
We've to write a couple of KeyCloak SPI's - Authentication and UserStorage - is this something that can we can do with quarkus or should it be a standard java project?

**Comment by codespearhead** [comment]
Plugin are meant to only use what's offered by bindings in the main application, which in Keycloak's case are SPIs, since those plugins will be bundled with Keycloak itself.

Hence, I'm not sure I understand your question.

This discussion might help: https://github.com/keycloak/keycloak/discussions/29243

**Reply by akil-rails** [reply]
I will clarify the question - currently for the keycloak SPI we use a standard java project , I was wondering if we can build the SPI as a Quarkus extension instead of a standard java project - with the extension we get lots of extra goodies like RestClient & Config amongst others -

**Reply by codespearhead** [reply]
Yes, you can use Quarkus or anything that compiles to Java bytecode to develop a Keycloak extension.

If your project is structured as a Quarkus extension and that Quarkus extension is meant to be deployed as an extension to another Quarkus instance running inside the Keycloak plugin, as opposed to an extension to the Quarkus instance that powers Keycloak, it'll work.

[p2-inc/keycloak-orgs](https://github.com/p2-inc/keycloak-orgs) is an example of a non-trivial plugin that, although doesn't use Quarkus, spins up another server with persistence and REST endpoints.

**Reply by arrudeboy** [reply]
Hi there! I am stucked with the same doubt. I am building a keycloak provider for RHBK and I would like to use jakarta **@Inject** annotation in order to inject some beans which are treated as singleton. Example:

```
@ApplicationScoped
public class ConsentAccountsRequiredAction implements RequiredActionProvider, RequiredActionFactory {

    @Inject
    AccountService accountService;

    @Override
    public void requiredActionChallenge(RequiredActionContext context) {
        var user = context.getUser();
        var customerId = user.getFirstAttribute("customerId");
        var accounts = accountService.getCbuList(customerId);
        ...
    }
```


But **accountService** object always remains null. 

I know that there're things that I am doing wrong. I would appreciate if anyone can provide an example of configuring beans within a SPI (in my case is a RequiredAction provider). 

Thanks in advance!

**Reply by codespearhead** [reply]
You asked the same question in [1]. I answered it there.

[1] https://github.com/keycloak/keycloak/discussions/12158

**Reply by sberyozkin** [reply]
Thanks @codespearhead, indeed it would benefit other Keycloak users as well to continue the conversation in the Keycloak discussion

**Reply by codespearhead** [reply]
@sberyozkin Are you able to transfer this discussion [1] to the Keycloak repository?

If you can mark my second-last comment as the accepted answer [2] for this discussion, so it no longer appears as unanswered in the GitHub discussions tab, even better.


***

[1] https://docs.github.com/en/discussions/managing-discussions-for-your-community/managing-discussions#transferring-a-discussion
[2] https://docs.github.com/en/discussions/managing-discussions-for-your-community/moderating-discussions#marking-a-comment-as-an-answer

**Reply by sberyozkin** [reply]
@codespearhead I can't as apparently one can move it to the repository within organization only. I'll mark you last comment above as an answer

**Reply by sberyozkin** [reply]
@codespearhead I think we can just close it here then, as I can't mark the comment as an answer either.
@akil-rails, if you'd like, please continue at the Keycloak discussions linked to by @codespearhead, thanks
